cmd/kubelet/app/server.go

```diff
@@ -377,116 +347,63 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
-			if err == nil {
-				// Update s (KubeletServer) with new config from API server
-				s.KubeletConfiguration = *remoteKC
-				// Ensure that /configz is up to date with the new config
-				if cfgzErr != nil {
-					glog.Errorf("was unable to register configz before due to %s, will not be able to set now", cfgzErr)
-				} else {
-					setConfigz(cfgz, &s.KubeletConfiguration)
-				}
-				// Update feature gates from the new config
-				err = utilfeature.DefaultFeatureGate.Set(s.KubeletConfiguration.FeatureGates)
-				if err != nil {
-					return err
-				}
-			} else {
-				glog.Errorf("failed to init dynamic Kubelet configuration sync: %v", err)
-			}
-		}
-	}
-
-	// Validate configuration.
-	if err := validateConfig(s); err != nil {
-		return err
-	}
-
+	var nodeName types.NodeName
 	if kubeDeps == nil {
-		var kubeClient clientset.Interface
-		var eventClient v1core.EventsGetter
-		var externalKubeClient clientgoclientset.Interface
-		var cloud cloudprovider.Interface
-
-		if !cloudprovider.IsExternal(s.CloudProvider) && s.CloudProvider != componentconfigv1alpha1.AutoDetectCloudProvider {
-			cloud, err = cloudprovider.InitCloudProvider(s.CloudProvider, s.CloudConfigFile)
-			if err != nil {
-				return err
-			}
-			if cloud == nil {
-				glog.V(2).Infof("No cloud provider specified: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
-			} else {
-				glog.V(2).Infof("Successfully initialized cloud provider: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
-			}
+		cloud, err := getCloud(s)
+		if err != nil {
+			return err
 		}
-
-		nodeName, err := getNodeName(cloud, nodeutil.GetHostname(s.HostnameOverride))
+		nodeName, err = getNodeName(cloud, nodeutil.GetHostname(s.HostnameOverride))
 		if err != nil {
 			return err
 		}
-
 		if s.BootstrapKubeconfig != "" {
 			if err := bootstrapClientCert(s.KubeConfig.Value(), s.BootstrapKubeconfig, s.CertDirectory, nodeName); err != nil {
 				return err
 			}
 		}
-
-		clientConfig, err := CreateAPIServerClientConfig(s)
-		if err == nil {
-			kubeClient, err = clientset.NewForConfig(clientConfig)
-			if err != nil {
-				glog.Warningf("New kubeClient from clientConfig error: %v", err)
-			}
-			externalKubeClient, err = clientgoclientset.NewForConfig(clientConfig)
-			if err != nil {
-				glog.Warningf("New kubeClient from clientConfig error: %v", err)
-			}
-			// make a separate client for events
-			eventClientConfig := *clientConfig
-			eventClientConfig.QPS = float32(s.EventRecordQPS)
-			eventClientConfig.Burst = int(s.EventBurst)
-			eventClient, err = clientgoclientset.NewForConfig(&eventClientConfig)
-			if err != nil {
-				glog.Warningf("Failed to create API Server client: %v", err)
-			}
-		} else {
-			if s.RequireKubeConfig {
-				return fmt.Errorf("invalid kubeconfig: %v", err)
-			}
-			if standaloneMode {
-				glog.Warningf("No API client: %v", err)
-			}
+		kubeClient, _, _, err := getClients(s, standaloneMode)
```

Comment:
@deads2k 
>1. refactor kubelet to run from a config file
>2. create a new process to manage the config file like

This PR is just one step on that journey. The `nodeconfig` controller is the thing that does both the loading and checkpointing of configurations. Eventually we'll be able to focus on just running from a file or a checkpoint, but almost all the existing flags need to be deprecated, configuration structures finalized, etc. before we can get there, so it's a work-in-progress.

@liggitt I hate the KubeletServer model and I completely agree wrt avoiding global state and having a logical, simple series of bootstrapping steps, with clearly defined i/o boundaries. I'm trying to eventually get us there. But unfortunately these difficult to reason about patterns exist all over the Kubelet and we're not going to fix all of them in this PR. I personally don't think it's appropriate for the PR that implements the dynamic config mechanism to also refactor the entire bootstrap sequence of the Kubelet; it is but one component of that sequence. I have ideas for what makes sense for a bootstrapping flow, but given the risk of derailing this PR I'm not going to discuss them here.

Fully replacing flags with config files is more like the last step, because we need the machinery in place for the new config paradigm before we can mark all the flags deprecated, and then we still have to wait out the deprecation period to rip them out. So the timeline looks more like: 
- 1.7 - alpha dynamic config mechanism implemented (which includes init config, which is "running the kubelet from a file")
- 1.8 kubelet configuration APIs are refactored and moved to beta (this consists of auditing the fields and determining what is safe to change dynamically and what isn't, providing a way for non-dynamic parameters to exist in the init config, etc.)
- 1.9 dynamic config can go GA and we can mark all the unnecessary flags deprecated
- 1.10 we can finally remove the unnecessary flags

It's not a short journey to get to a clean bootstrap flow, but we're working towards it one step at a time.